python/lsst/obs/base/gen2to3/translators.py

```diff
@@ -276,55 +278,41 @@ def extract(self, gen2id, *args, **kwargs):
-
-    Translators should usually be constructed via the `makeMatching` method.
-
-    Parameters
-    ----------
-    handlers : `list`
-        A list of KeyHandlers this Translator should use.
-    skyMap : `BaseSkyMap`, optional
-        SkyMap instance used to define any tract or patch Dimensions.
-    skyMapName : `str`
-        Gen3 SkyMap Dimension name to be associated with any tract or patch
-        Dimensions.
-    datasetTypeName : `str`
-        Name of the dataset type whose data IDs this translator handles.
+class TranslatorFactory:
+    """A class that manages a system of rules for translating Gen2 data IDs
+    to Gen3 data IDs, and uses these to construct translators for particular
+    dataset types.
     """
-    def __init__(self, handlers: List[KeyHandler], skyMap: Optional[BaseSkyMap], skyMapName: Optional[str],
-                 datasetTypeName: str):
-        self.handlers = handlers
-        self.skyMap = skyMap
-        self.skyMapName = skyMapName
-        self.datasetTypeName = datasetTypeName
-
-    __slots__ = ("handlers", "skyMap", "skyMapName", "datasetTypeName")
-
-    # Rules used to match Handlers when constring a Translator.
-    # outer key is instrument name, or None for any
-    # inner key is DatasetType name, or None for any
-    # values are 3-tuples of (frozenset(gen2keys), handler, consume)
-    _rules: Dict[
-        Optional[str],
-        Dict[
-            Optional[str],
-            Tuple[FrozenSet[str], KeyHandler, bool]
-        ]
-    ] = {
-        None: {
-            None: []
+    def __init__(self):
+        # The rules used to match KeyHandlers when constructing a Translator.
+        self._rules: Dict[
+            Optional[str],  # instrument name (or None to match any)
+            Dict[
+                Optional[str],  # dataset type name (or None to match any)
+                # gen2keys, handler, consume
+                Tuple[FrozenSet[str], KeyHandler, bool]
+            ]
+        ] = {
+            None: {
+                None: []
+            }
         }
-    }
+        self._addDefaultRules()
 
     def __str__(self):
-        hstr = ",".join(str(h) for h in self.handlers)
-        return f"{type(self).__name__}(dtype={self.datasetTypeName}, handlers=[{hstr}])"
-
-    @classmethod
-    def addRule(cls, handler: KeyHandler, instrument: Optional[str] = None,
+        lines = []
+        for instrumentName, nested in self._rules.items():
+            if instrumentName is None:
+                instrumentName = "[any instrument]"
+            for datasetTypeName, (gen2keys, handler, consume) in nested.items():
+                if datasetTypeName is None:
+                    datasetTypeName = "[any dataset type]"
+                consumed = " (consumed)" if consume else ""
+                lines.append(f"{instrumentName} + {datasetTypeName}, for data ID "
+                             f"with {gen2keys}{consumed}: {handler}.")
+        return "\n".join(lines)
```

Comment:
Given that this is a somewhat complicated `str`, it would be good to have a few tests for it with the different bits being `None`. I think that can easily go in `test_instrument.py`, since the DummyCam there has a default Factory.

Reply:
I'm happy to test that it doesn't raise due to a bug since that's _super_ annoying, but I think this is a convenience for debugging, and I don't think testing content is worth the maintenance/fragility inherent in those kinds of stringification tests here.

Reply:
At least a basic test that you get out a reasonable string for a reasonable Factory.

Reply:
Now I'm sort of regretting adding the `__str__`, because the effort I've put into it has now probably exceeded the value I think we'd get out of it.  Jenkins is running with a trivial doesn't-raise test, and I'm hoping I can still make the weekly with that.  When we talk about testing philosophy someday at a pipelines meeting, we should make sure we talk about testing stringification, because I think there are a lot of cases where the value added by tests for those is actually negative.  Stringification for humans usually defines the kind of stable interface that tests should cover (though the best kind of `__repr__` is a notable exception).